rfc/rfc-37/rfc-37.md

```diff
@@ -0,0 +1,286 @@
+<img src="metadata_index_col_stats.png" alt="Column Stats Partition" width="480"/>
+
+### Metadata Index lookup:
+
+For the incoming upsert records, given their keys, tag their current location. The new algorithm for the
+index lookup would be
+
+1. Generate the list of partitions and the list of keys under each partition to be looked up
+2. For all the involved partitions, load all its file list
+3. Level 1: Range pruning using `column_stats` index:
+   1. For each of the record key, generate the column stats index lookup key based on the tuple 
+      (__hoodie_record_key, partition name, file path)
+   2. Meta index lookup with the above key and if available get the value payload with the column stats details
+   3. Build the key range interval tree from all the looked up keys
+   4. Prune the partition and its candidate files based on the range interval lookups
+4. Leve 2: Record pruning using `bloom_filter`  index:
+   1. From the shortlisted file candidates per partition, generate bloom filter index lookup key based on the tuple
+      (partition name, file path)
+   2. Meta index lookup with the above key to load the base file bloom filter
+   3. Bloom filter lookup for the record key to generate the probable keys
+5. Level 3: Record validation
+   1. Given the list of files and their probable keys from above pruning, do the actual file lookup to
+      confirm the keys
+   2. Return the location (file id) of the final matching keys
+
+### Schema Evolution:
+
+HashID based key are deterministically generated from the tuple input. Mean, for the tuple consisting of column name,
+partition name and file name, the key generated would always be the same. So, a table where the schema gets changed over
+time would have an impact on the keys already generated. The most common schema evolution use cases like change of
+column type, adding a new column are not affected though. Other relatively uncommon use cases like column name rename,
+dropping a column and adding a column with dropped name would have indices referring them more than needed. This would
+lead to the index lookup matching stale/new records across evolved schemas. 
+
+To avoid looking up stale/new index records, here are the design options we have:
+1. (Preferred) Query rewrite / Result recordset pruning
+   1. Schema evolution layer should introduce query rewrite stage to detect evolved schemas for the input query and
+      optionally include additional predicates to the query
+   2. The resultant recordset can also be pruned based on the commit time and the schema change time
+3. Making input tuple set schema aware 
+   1. Along with column name, partition name and file path, a version/tag can also be added to make the key
+      generated very schema specific. But, this choice has a performance impact as the lookup now has to be more of a
+      prefix based instead of pointed lookups. That is, index lookup have to return records for all the versions/tags
```

Comment:
Not sure i understand why this is the case: my understanding is that we're only aiming for correctness (meaning that we want to make sure that schema evolution doesn't lead to incorrect results) rather than planning on time-traveling support across schema-change boundaries. In that case we only need to incorporate schema version in our key (for proper attribution) and on the query path we always use the latest schema version, hence if schema changed b/w the time we've indexed and the query time, query simply will miss on the index and will have to fallback to a no-index mode.

Reply:
I guess we are not looking for any fallbacks here. we want the col stats to return the right results always even considering schema evolved. We can discuss more if you are interested to deep dive on this regard.

Reply:
How would that work? For example, if i would upgrade and only after a while enable the property guarding this i won't have indexes for all the files i created.

Similar case would be if the schema evolves: if the writes comes with a new schema none of the existing files would have it.

Reply:
Schema versions and evolution logic are not pushed to the index layer. The layer above have to modify the query based on its knowledge of when the evolution and what happened. Mean, the index is going to return all possible results for the key being lookedup and the layer above have to do any pruning if needed. 

Regarding index not available, yes we are taking care of this in the async index building project. Not all index will be available and ready for use all the times and we have to fallback if not available.

Reply:
@manojpec i think we need to be more specific than just saying that the layer above will handle that: how would it do that? How is it going to rewrite query? 

What we primarily aim for here is correctness, therefore we have to guarantee that even in the presence of schema evolution index will continue to return correct results. 

> 3. Making input tuple set schema aware 
>    1. Along with column name, partition name and file path, a version/tag can also be added to make the key
>       generated very schema specific. But, this choice has a performance impact as the lookup now has to be more of a
>       prefix based instead of pointed lookups. That is, index lookup have to return records for all the versions/tags

This would be a penalty only on the fields evolved proportional to the number of times particular field has evolved. That doesn't seem like a considerable perf penalty.

On top of that we can order keys in descending order making version tags padded w/ 0s like following
```
003
002
001
```

Which will guarantee that the prefix lookup of the first matching record will always yield us latest version.

Reply:
Sync'd up offline w/ @manojpec on this. 

TL;DR is that current MT design already handling all of the use-cases for schema evolution (column deletion/addition, column renames) so there's actually no additional support required.